File: backend/src/main/java/com/backend/ord/domain/entities/LanguageProficiency.java
package com.backend.ord.domain.entities;

import com.backend.ord.domain.entities.abstracts.EntityBase;
import com.backend.ord.enums.Language.LanguageName;
import com.backend.ord.enums.Language.LanguageProficiencyLevel;
import jakarta.persistence.*;

@Entity
@Table(name = "language_proficiencies")
public class LanguageProficiency extends EntityBase {
    @Column(name = "language", nullable = false)
    @Enumerated(EnumType.STRING)
    private LanguageName language;

    @Column(name = "proficiency", nullable = false)
    @Enumerated(EnumType.STRING)
    private LanguageProficiencyLevel proficiency;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", referencedColumnName = "id")
    private User user;

    public LanguageProficiency(LanguageName language, LanguageProficiencyLevel proficiency, User user) {
        this.language = language;
        this.proficiency = proficiency;
        this.user = user;
    }

    public LanguageProficiency() {
    }

    public static LanguageProficiencyBuilder builder() {
        return new LanguageProficiencyBuilder();
    }

    public LanguageName getLanguage() {
        return this.language;
    }

    public LanguageProficiencyLevel getProficiency() {
        return this.proficiency;
    }

    public User getUser() {
        return this.user;
    }

    public void setLanguage(LanguageName language) {
        this.language = language;
    }

    public void setProficiency(LanguageProficiencyLevel proficiency) {
        this.proficiency = proficiency;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public static class LanguageProficiencyBuilder {
        private LanguageName language;
        private LanguageProficiencyLevel proficiency;
        private User user;

        LanguageProficiencyBuilder() {
        }

        public LanguageProficiencyBuilder language(LanguageName language) {
            this.language = language;
            return this;
        }

        public LanguageProficiencyBuilder proficiency(LanguageProficiencyLevel proficiency) {
            this.proficiency = proficiency;
            return this;
        }

        public LanguageProficiencyBuilder user(User user) {
            this.user = user;
            return this;
        }

        public LanguageProficiency build() {
            return new LanguageProficiency(this.language, this.proficiency, this.user);
        }

        public String toString() {
            return "LanguageProficiency.LanguageProficiencyBuilder(language=" + this.language + ", proficiency=" + this.proficiency + ", user=" + this.user + ")";
        }
    }
}

File: backend/src/main/kotlin/com/backend/ord/domain/persistence/entities/LanguageProficiency.kt
package com.backend.ord.domain.persistence.entities

import com.backend.ord.domain.persistence.entities.interfaces.IdentifiableUserResource
import com.backend.ord.enums.persistence.language.LanguageName
import com.backend.ord.enums.persistence.language.LanguageProficiencyLevel
import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.*

@Entity
@Table(name = "language_proficiencies")
data class LanguageProficiency(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    override var id: UUID = UUID.randomUUID(),

    @Column(name = "language", nullable = false)
    @Enumerated(EnumType.STRING)
    var language: LanguageName,

    @Column(name = "proficiency", nullable = false)
    @Enumerated(EnumType.STRING)
    var proficiency: LanguageProficiencyLevel,

    @Column(name = "generative_content_language", nullable = false)
    @Enumerated(EnumType.STRING)
    var generativeContentLanguage: LanguageName,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", referencedColumnName = "id")
    override var user: User,

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    @UpdateTimestamp
    var updatedAt: Instant = Instant.now()
) : IdentifiableUserResource
